### PR TITLE
`find_domain` for `astype`

### PR DESCRIPTION
**Pull request by ordabayevy**
Addresses #560

### CODE REVIEW

**Review by eb8680 — CHANGES_REQUESTED** [review]
Can you add a test showing this does indeed resolve #560?

File: test/test_cnf.py [hunk 1]
```diff
@@ -130,3 +134,21 @@ def test_eager_contract_tensor_tensor(
...
+    integrand = Constant((("x", Real),), data)
+    with reflect:
+        term = Integrate(log_measure, integrand, "x")
+        result = term.reduce(ops.add, {"i", "j"})
+
+    with eager:
+        expected = reinterpret(result)
+
+    actual = apply_optimizer(result)
+    assert_close(actual, expected)
```
Comment (ordabayevy) [inline test/test_cnf.py:154]: This is as much as I could simplify the test that would still fail.

**Review by eb8680 — APPROVED** [review]
Looks good

**Comment by ordabayevy** [comment]
Thanks, @eb8680 !

**Pull request merged**